xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/listeners/MentionsCreatedEventListener.java

```diff
@@ -74,16 +76,28 @@ public void onEvent(Event event, Object source, Object data)
             DocumentCreatedEvent.class.getName(), source, data);
 
         XWikiDocument doc = (XWikiDocument) source;
-        MentionsCreatedRequest mentionsCreatedRequest = new MentionsCreatedRequest(doc);
-        mentionsCreatedRequest.setVerbose(false);
+        DocumentReference authorReference = doc.getAuthorReference();
+        DocumentReference documentReference = doc.getDocumentReference();
 
-        try {
-            this.jobExecutor.execute(MentionsCreateJob.ASYNC_REQUEST_TYPE, mentionsCreatedRequest);
-        } catch (JobException e) {
-            this.logger.warn(
-                "Failed to create a Job for the Event [{}] received from [{}] with data [{}]. Cause: [{}]",
-                DocumentCreatedEvent.class.getName(), source, data, getRootCauseMessage(e));
-        }
+        this.executor.executeCreate(doc.getXDOM(), authorReference, documentReference, DOCUMENT);
+
+        traverseXObjects(doc.getXObjects(), authorReference, documentReference);
     }
 
+    private void traverseXObjects(Map<DocumentReference, List<BaseObject>> xObjects, DocumentReference authorReference,
+        DocumentReference documentReference)
+    {
+        for (Map.Entry<DocumentReference, List<BaseObject>> entry : xObjects.entrySet()) {
+            for (BaseObject baseObject : entry.getValue()) {
+                if (baseObject != null) {
+                    for (Object o : baseObject.getProperties()) {
+                        if (o instanceof LargeStringProperty) {
+                            String content = ((LargeStringProperty) o).getValue();
+                            this.executor.executeCreate(content, authorReference, documentReference, AWM_FIELD);
```

Comment:
I'm wondering if putting all that in the queue is not a bit expensive in term of memory (imagine what will end up in the queue when we import a big backup).

Might be safer to just queue the reference of documents to process and then get the document and its previous version in the background thread. I guess it should not increase too much the processing time if the mentions found in the previous version of the document are cached.